polyglot/piranha/src/models/source_code_unit.rs

```diff
@@ -72,12 +80,59 @@ impl SourceCodeUnit {
 
   pub(crate) fn apply_edit(&mut self, edit: &Edit, parser: &mut Parser) -> InputEdit {
     // Get the tree_sitter's input edit representation
-    self._apply_edit(
+    let mut applied_edit = self._apply_edit(
       edit.replacement_range(),
       edit.replacement_string(),
       parser,
       true,
-    )
+    );
+    // Check if the edit kind is "DELETE something"
+    if edit.replacement_string().is_empty() {
+      let deleted_at = edit.replacement_range().start_point.row;
+      if let Some(comment_range) = self.get_comment_at_line(deleted_at, 2) {
+        info!("Deleting an associated comment");
+        applied_edit = self._apply_edit(comment_range, "", parser, false);
+      }
+    }
+    return applied_edit;
+  }
+
+  /// This function reports the range of the comment associated to the deleted element. 
+  /// 
+  /// # Arguments:
+  /// * row : The row number where the deleted element started
+  /// * buffer: Number of lines that we want to look up to find associated comment 
+  /// 
+  /// # Algorithm : 
+  /// Get all the nodes that either start and end at [row]
+  /// If **all** nodes are comments 
+  /// * return the range of the comment
+  /// If the [row] has no node that either starts/ends there:
+  /// * recursively call this method for [row] -1 (until buffer is positive)
+  fn get_comment_at_line(&mut self, row: usize, buffer: usize) -> Option<Range> {
+    // Get all nodes that start or end on `updated_row`.
+    let mut relevant_nodes_found = false;
+    let mut relevant_nodes_are_comments = true;
+    let mut comment_range = None;
+    for node in traverse(self.ast.walk(), Order::Post) {
```

Comment:
Is `self.ast` the full file? This looks like a linear scan over the ast, don't we already have the position of the deleted node somewhere? This seems relatively inefficient to perform for every deletion, at least on large refactorings (probably not a problem for XP flags, since a lot of code is scanned but only a few files have deletions...)

Reply:
Yes. This is expensive. But we can turn this off or limit its impact now (by disabling/ selectively disabling) comment cleanup.

Reply:
Is there a way to get the parent of the AST element that just got deleted? If we have that, we can scan for comments only there, there is no point to scan the whole file AST...